Add option to delete user profile in Profile Selector (adds #38243)

JuhoErvasti opened the pull request:
This is my implementation of this somewhat old feature request #38243 to be able to remove user profiles in the GUI without having to manually delete the folder.

* I figured that a logical place to add this feature is the recently added QgsUserProfileSelectionDialog.

![](https://github.com/user-attachments/assets/911702bc-e708-421a-a7ab-ca4637ed74b0)

* Since not everyone has the selection dialog enabled I thought I'd add a button to open the Selector from **Settings** > **User Profiles**

![image](https://github.com/user-attachments/assets/17f452df-b334-400a-8920-1e42c8b77d5b)

* I've also disabled deleting the default profile, although I'm not sure it's necessary

---
If this is at all an okay direction for the Profile Selector, personally I'd like for it to have even more features:
* Being able to select and delete multiple profiles
* Being able to add profiles without immediately starting a new process
* Being able to rename profiles
* Maybe being able to reset a profile?
* Maybe being able to edit the profile icons?

JuhoErvasti committed:
Add minimal version of profile deletion from GUI

JuhoErvasti committed:
Improve profile deletion

- Ask to confirm deletion
- Refresh list of profiles after deletion
- Disallow deleting default profile (not sure if necessary?)
- Check for errors

JuhoErvasti committed:
Change confirmation to QMessageBox::Yes and No

- I figure this is better and safer since with Ok and Cancel the default
  option is Ok so a user might accidentally confirm deletion with the
  keyboard, now the default option is No

JuhoErvasti committed:
Enable opening profile selector from Profiles menu

JuhoErvasti committed:
Move open profile selector action to the bottom

JuhoErvasti committed:
Strengthen warning about deleting the profile

Gustry commented on src/app/qgisapp.cpp:
```diff
@@ -3544,6 +3545,19 @@ void QgisApp::refreshProfileMenu()
+
+  mConfigMenu->addSeparator( );
+
+  QAction *openProfileSelectionDialog = mConfigMenu->addAction( tr( "Open Profile Selector" ) );
```
```suggestion
  QAction *openProfileSelectionDialog = mConfigMenu->addAction( tr( "Open Profile Selector…" ) );
```

As a dialog will be opened after the click and the user will need to interact with it.

JuhoErvasti replied:
Good point! Added in 5eb0a50442444acb8cfb542bb97a904badcfeabc

3nids commented:
This is super nice. I would be in favor of adding this to 3.40. Thoughts?

JuhoErvasti committed:
Gray out and disable choosing active profile

- I think this is clearer and also matches the behaviour of the User
  Profiles menu
- Also add ... to indicate a dialog

JuhoErvasti commented:
![image](https://github.com/user-attachments/assets/254d2c70-374e-409e-b918-e9b706acf5ff)
I fixed some oversights

* When opening the dialog from the GUI the active profile is greyed out and cannot be opened **or removed**
  * I thought this would be good since it matches the behavior of the User Profiles menu
* Also as per Gustry's review the action now has ... to indicate that you're opening a dialog

> This is super nice. I would be in favor of adding this to 3.40. Thoughts?

For my part I'd like that if it's possible.

nicogodet commented on src/app/options/qgsuserprofileselectiondialog.cpp:
```diff
@@ -104,4 +116,36 @@ void QgsUserProfileSelectionDialog::onAddProfile()
+
+  if ( selectedProfile == QStringLiteral( "default" ) )
+  {
+    QMessageBox::warning( this, tr( "Remove profile" ), QString( "Cannot delete default profile!" ), QMessageBox::Ok );
```
```suggestion
    QMessageBox::warning( this, tr( "Remove profile" ), tr( "Cannot delete default profile!" ), QMessageBox::Ok );
```

JuhoErvasti replied:
My bad, all fixed in cb80c740e4c631c22ff7b2c84f1c3fbbbf09204d

nicogodet commented on src/app/options/qgsuserprofileselectiondialog.cpp:
```diff
@@ -104,4 +116,36 @@ void QgsUserProfileSelectionDialog::onAddProfile()
+
+  if ( selectedProfile == mActiveProfile )
+  {
+    QMessageBox::warning( this, tr( "Remove profile" ), QString( "Cannot delete active profile!" ), QMessageBox::Ok );
```
```suggestion
    QMessageBox::warning( this, tr( "Remove profile" ), tr( "Cannot delete active profile!" ), QMessageBox::Ok );
```

nicogodet commented on src/app/options/qgsuserprofileselectiondialog.cpp:
```diff
@@ -104,4 +116,36 @@ void QgsUserProfileSelectionDialog::onAddProfile()
+    return;
+  }
+
+  QMessageBox::StandardButton response = QMessageBox::warning( this, tr( "Remove profile" ), QString( "Are you sure you want to delete profile '%1'? This action cannot be undone!" ).arg( selectedProfile ), QMessageBox::Yes | QMessageBox::No );
```
```suggestion
  QMessageBox::StandardButton response = QMessageBox::warning( this, tr( "Remove profile" ), tr( "Are you sure you want to delete profile '%1'? This action cannot be undone!" ).arg( selectedProfile ), QMessageBox::Yes | QMessageBox::No );
```

nyalldawson commented on src/app/options/qgsuserprofileselectiondialog.cpp:
```diff
@@ -37,14 +38,25 @@ QgsUserProfileSelectionDialog::QgsUserProfileSelectionDialog( QgsUserProfileMana
   {
     auto item = new QListWidgetItem( mManager->profileForName( profile )->icon(), profile );
+    if ( profile == mActiveProfile )
+      item->setBackground( Qt::lightGray );
```
This won't play nice with systems using dark themes -- we need to instead get a color from the widget's palette().

nyalldawson commented on src/app/options/qgsuserprofileselectiondialog.cpp:
```diff
@@ -104,4 +116,36 @@ void QgsUserProfileSelectionDialog::onAddProfile()
+
+  if ( selectedProfile == QStringLiteral( "default" ) )
+  {
+    QMessageBox::warning( this, tr( "Remove profile" ), tr( "Cannot delete default profile!" ), QMessageBox::Ok );
```
```suggestion
    QMessageBox::warning( this, tr( "Remove Profile" ), tr( "The default profile cannot be deleted." ), QMessageBox::Ok );
```

nyalldawson commented on src/app/options/qgsuserprofileselectiondialog.cpp:
```diff
@@ -104,4 +116,36 @@ void QgsUserProfileSelectionDialog::onAddProfile()
+
+  if ( selectedProfile == mActiveProfile )
+  {
+    QMessageBox::warning( this, tr( "Remove profile" ), tr( "Cannot delete active profile!" ), QMessageBox::Ok );
```
```suggestion
    QMessageBox::warning( this, tr( "Remove Profile" ), tr( "The active profile cannot be deleted." ), QMessageBox::Ok );
```

nyalldawson commented on src/app/options/qgsuserprofileselectiondialog.cpp:
```diff
@@ -104,4 +116,36 @@ void QgsUserProfileSelectionDialog::onAddProfile()
+    return;
+  }
+
+  QMessageBox::StandardButton response = QMessageBox::warning( this, tr( "Remove profile" ), tr( "Are you sure you want to delete profile '%1'? This action cannot be undone!" ).arg( selectedProfile ), QMessageBox::Yes | QMessageBox::No );
```
```suggestion
  QMessageBox::StandardButton response = QMessageBox::warning( this, tr( "Remove Profile" ), tr( "Are you sure you want to delete profile '%1'? This action cannot be undone!" ).arg( selectedProfile ), QMessageBox::Yes | QMessageBox::No );
```

nyalldawson commented on src/app/options/qgsuserprofileselectiondialog.cpp:
```diff
@@ -104,4 +116,36 @@ void QgsUserProfileSelectionDialog::onAddProfile()
+  }
+  else
+  {
+    QMessageBox::warning( this, tr( "Remove profile" ), tr( "Profile '%1' could not be deleted!" ).arg( selectedProfile ) );
```
```suggestion
    QMessageBox::warning( this, tr( "Remove Profile" ), tr( "Profile '%1' could not be deleted!" ).arg( selectedProfile ) );
```

nyalldawson commented on src/app/qgisapp.cpp:
```diff
@@ -3544,6 +3545,20 @@ void QgisApp::refreshProfileMenu()
+  openProfileSelectionDialog->setObjectName( "mActionOpenProfileSelector" );
+  connect( openProfileSelectionDialog, &QAction::triggered, this, [this, activeName]()
+  {
+    auto dlg = QgsUserProfileSelectionDialog( userProfileManager(), activeName );
+
+    if ( dlg.exec() == QDialog::Accepted )
+    {
+      userProfileManager()->loadUserProfile( dlg.selectedProfileName() );
+    }
```
QGIS coding convention is to avoid `auto` for anything but iterator/heavily template based types:

```suggestion
    QgsUserProfileSelectionDialog dlg( userProfileManager(), activeName );

    if ( dlg.exec() == QDialog::Accepted )
    {
      userProfileManager()->loadUserProfile( dlg.selectedProfileName() );
    }
```

JuhoErvasti replied:
My bad, fixed.

nyalldawson commented:
I'd suggest we re-word "remove profile" to "**delete** profile" everywhere, as it's clearer that it's a destructive action rather then just removing it from the list. (We use "remove" elsewhere for actions like removing a connection to a geopackage, where the file won't be actually be deleted).

JuhoErvasti commented:
> I'd suggest we re-word "remove profile" to "delete profile" everywhere, as it's clearer that it's a destructive action

That makes a lot of sense, it was very inconsistent too. I've changed the term to 'delete' across the board here 3658792c1e693762b3a1dff726a310edbcd17d20

JuhoErvasti committed:
Address review

DelazJ commented:
Thanks @JuhoErvasti for this feature
> I've also disabled deleting the default profile, although I'm not sure it's necessary

I do not see why this is necessary. The so-called "default" profile can be corrupted after some configuration and it could be nice to get rid of it. Afaict, if there is no profile, QGIS will create a new clean one, so we should just be careful not deleting the one in use. BTW (probably an edge case but) is it the one in use or the oneS in use that should be preserved from deletion?

JuhoErvasti commented:
> I do not see why this is necessary.

I think I agree that you should be able to delete the default profile too, since as you say a new one is created anyways.

> BTW (probably an edge case but) is it the one in use or the oneS in use that should be preserved from deletion?

This is a good point. I think ideally it _should_ check **all** active profiles, since deleting an active profile can cause problems. Although AFAIK there is currently no way to check which profiles are open in every different process.

nicogodet commented:
My two cents, _default_ profile might not even exists in certain circumstances: if you start QGIS with `qgis --profile my_profile` on fresh install, profile named _default_ will not exists.

JuhoErvasti committed:
Re-enable deleting profile named 'default'

JuhoErvasti commented:
I've enabled deleting the default profile (or the profile _named_ default, to be more accurate I suppose), are there any further changes I should make?

DelazJ commented:
@JuhoErvasti I can't speak code-wise but I tested the artifact and it works nicely, as described.
> > BTW (probably an edge case but) is it the one in use or the oneS in use that should be preserved from deletion?

> This is a good point. I think ideally it should check all active profiles, since deleting an active profile can cause problems. Although AFAIK there is currently no way to check which profiles are open in every different process.

replying to self: I tested, and it is not possible to delete an opened profile from another opened profile. But it is not done gracefully because you are asked if you really want to do it, yes, and then you get the no-go (probably because there is a lock somewhere). We should better directly tell people that it isn't removable (and why).
> I've enabled deleting the default profile (or the profile named default, to be more accurate I suppose)

Your comment made me remind that in QGIS we have a profile named "default" and we also have a "defaultProfile". you can find the latter in in the `profiles.ini` file (defined in Settings --> Options --> User profiles). The first is the one we were discussing about above and we agree it is removable.
My question is "should we also allow removal of the other one, the one that is/can be set as "always use profile"? I opted to "always use a profile" and could delete it from another one profile; it was silently replaced in the config by the first (alphabetically sorted) profile I had. Also happens if I select "use last profile". Not an issue from your PR but wanted to share and ask for opinions.

JuhoErvasti commented:
@DelazJ Regarding the deletion of opened profiles: I tested it from the artifact as well and it looks like it's a bit more complicated. The profile selector does report that the profile couldn't be deleted, but in reality everything except some .db files **are in fact deleted**. When you close the process with the "unsuccessfully" deleted profile then it's saved a bit like a fresh profile, almost everything was deleted.

I think this is pretty misleading, and an oversight on my part - sorry for that. IMO this shouldn't be merged before this gets addressed. I think it does need to be checked that the profiled selected for deletion is not opened anywhere.

I'll try to look into this more but it might take some time. I'm open to any ideas, as far as I can tell solving this might be quite complicated.
